File: src/helmlab/metrics/confidence.py
"""Confidence layer for color-difference predictions (EXPERIMENTAL, v2).

A standard color-difference metric returns one number: how different two colors
are. This layer adds a second: how much *observers will disagree* about that
difference, predicted from the colors alone.

Provenance (honest, v2 — ordinal refit 2026-07-05)
---------------------------------------------------
- Base metric: helmlab's perceptual distance (v21 MetricSpace).
- HumanFB's ratings are 5-level CATEGORICAL; v1 used the std of the category
  codes as its target, which turned out to be ANTI-correlated (rho = -0.51)
  with the true latent observer spread (ceiling pile-up artifact). v2 fits an
  ordered-probit model instead: categories are thresholds ``tau`` on a latent
  difference axis; each pair gets a latent mean ``mu`` and observer-spread
  ``sigma`` (log-sigma ridge, converged).
- Runtime model: ``mu = mu_scale*de`` (Spearman 0.957 vs fitted latents) and
  ``sigma = a*de + b*chroma + c`` (LOO-R2 = 0.49, latent units).
- Relative disagreement sigma/mu falls with de (rho = -0.77) and chroma
  (-0.36): small / low-chroma differences are the unreliable regime.
- Cross-domain check (hong_2025, 8 observers): between-observer JND CV vs
  chroma-radius = +0.23, same sign as sigma's chroma term — but threshold and
  suprathreshold domains differ; do not over-read the chroma axis.

Limits
------
EXPERIMENTAL. n=47 training pairs, single uncontrolled collection session,
pairs sampled from helmlab's own zones. Calibrated for the small /
near-threshold regime (de <~ 0.15). Coefficients should not be over-read.
"""

import json
import os

import numpy as np

_PARAMS_PATH = os.path.join(os.path.dirname(__file__), "..", "data", "confidence_params.json")


def _load_params() -> dict:
    with open(_PARAMS_PATH) as f:
        return json.load(f)


def _erf(x):
    """Abramowitz & Stegun 7.1.26 (max err 1.5e-7) — IDENTICAL arithmetic to the
    JS sibling so cross-language parity holds; do not swap for math.erf."""
    x = np.asarray(x, float)
    sign = np.where(x < 0, -1.0, 1.0)
    ax = np.abs(x)
    t = 1.0 / (1.0 + 0.3275911 * ax)
    y = 1.0 - (((((1.061405429 * t - 1.453152027) * t) + 1.421413741) * t
                - 0.284496736) * t + 0.254829592) * t * np.exp(-ax * ax)
    return sign * y


def _phi(x):
    return 0.5 * (1.0 + _erf(np.asarray(x, float) / np.sqrt(2.0)))


class ConfidenceModel:
    """Predict observer disagreement (and a reliability score) for a color difference.

    Parameters
    ----------
    params : dict, optional
        Coefficients ``{a, b, c, scale, dis_floor, de_train_max, ...}``.
        Defaults to the bundled ``confidence_params.json``.
    """

    def __init__(self, params: dict | None = None):
        self.p = params if params is not None else _load_params()

    def disagreement(self, de, chroma):
        """Predicted inter-observer spread sigma (latent difference units)."""
        p = self.p
        d = p["sigma_a"] * np.asarray(de, float) + p["sigma_b"] * np.asarray(chroma, float) + p["sigma_c"]
        return np.maximum(d, p["sigma_floor"])

    def assess(self, de, chroma) -> dict:
        """Latent-model confidence for a perceptual difference ``de``.

        Returns ``de``, ``latent`` (mu), ``disagreement`` (sigma),
        ``reliability`` (mu/(mu+sigma)), ``p_noticeable``
        (P(latent > tau[1]) — chance a random observer rates the pair at
        least 'moderately different'), ``reliable`` (p_noticeable > 0.5)
        and ``extrapolated``.
        """
        de = np.asarray(de, float)
        p = self.p
        mu = p["mu_scale"] * de
        sig = self.disagreement(de, chroma)
        tau2 = p["tau"][1]
        p_not = 1.0 - _phi((tau2 - mu) / sig)
        out = {
            "de": de,
            "latent": mu,
            "disagreement": sig,
            "reliability": mu / (mu + sig),
            "p_noticeable": p_not,
            "reliable": mu > tau2,
            "extrapolated": de > p["de_train_max"],
        }
        if de.ndim == 0:
            out = {k: (float(v) if isinstance(v, np.ndarray) and v.ndim == 0
                       else (bool(v) if isinstance(v, (bool, np.bool_)) else v))
                   for k, v in out.items()}
        return out

File: packages/helmlab-js/src/core/confidence.ts
/**
 * Confidence layer for color-difference predictions (EXPERIMENTAL, v2).
 *
 * v2 = ordered-probit refit of HumanFB (2026-07-05): the 5-level categorical
 * ratings are modelled as thresholds `TAU` on a latent difference axis; each
 * pair gets a latent mean mu and observer-spread sigma. v1's target (std of
 * category codes) was ANTI-correlated (rho = -0.51) with the true latent
 * spread — a ceiling pile-up artifact. Runtime model: mu = MU_SCALE*de
 * (Spearman 0.957), sigma = A*de + B*chroma + C (LOO-R2 0.49, latent units).
 * Relative disagreement sigma/mu falls with de (-0.77) and chroma (-0.36):
 * small / low-chroma differences remain the unreliable regime. Mirrors
 * Python `helmlab.metrics.confidence` exactly (same coefficients, same erf).
 *
 * Limits: EXPERIMENTAL, n=47 pairs, zone-conditional sampling, de <~ 0.15.
 */
const MU_SCALE = 20.827694941963827;
const A = 5.785073154058929;
const B = 1.0148932574050995;
const C = -0.12266288972117914;
const SIGMA_FLOOR = 0.35;
const TAU2 = 0.7421522316160746;
const DE_TRAIN_MAX = 0.14811473587883495;

/** 1 JND in trained-metric de units: TAU2 / MU_SCALE — the de at which the
 * median observer of the ordinal model rates a pair at least "moderately
 * different". Used by hl.metric.jnd(); mirrors Python (tau[1]/mu_scale). */
export const JND_DE = TAU2 / MU_SCALE;
const SQRT2 = Math.sqrt(2.0);

/** Abramowitz & Stegun 7.1.26 — IDENTICAL arithmetic to the Python sibling
 *  (do not swap for a different erf; cross-language parity depends on it). */
function erf(x: number): number {
  const sign = x < 0 ? -1.0 : 1.0;
  const ax = Math.abs(x);
  const t = 1.0 / (1.0 + 0.3275911 * ax);
  const y = 1.0 - (((((1.061405429 * t - 1.453152027) * t) + 1.421413741) * t
              - 0.284496736) * t + 0.254829592) * t * Math.exp(-ax * ax);
  return sign * y;
}

const phi = (x: number): number => 0.5 * (1.0 + erf(x / SQRT2));

export interface Confidence {
  /** the perceptual distance itself */
  de: number;
  /** latent difference mu = MU_SCALE * de (ordinal-model units) */
  latent: number;
  /** predicted inter-observer spread sigma (latent units) */
  disagreement: number;
  /** mu / (mu + sigma), in [0, 1) */
  reliability: number;
  /** P(latent > TAU2) — chance a random observer rates the pair at least "moderately different" */
  pNoticeable: number;
  /** true when pNoticeable > 0.5 (mu > TAU2) */
  reliable: boolean;
  /** true if `de` is beyond the trained range (extrapolating) */
  extrapolated: boolean;
}

export function assessConfidence(de: number, chroma: number): Confidence {
  const mu = MU_SCALE * de;
  const sigma = Math.max(A * de + B * chroma + C, SIGMA_FLOOR);
  const pNoticeable = 1.0 - phi((TAU2 - mu) / sigma);
  return {
    de,
    latent: mu,
    disagreement: sigma,
    reliability: mu / (mu + sigma),
    pNoticeable,
    reliable: mu > TAU2,
    extrapolated: de > DE_TRAIN_MAX,
  };
}
